### ISSUE TITLE
Recommended padding for 5' Model

### ISSUE THREAD

**Issue by Mjvolk3** [gagneurlab/SpeciesLM#1]
In ModelUsage.py you have the following comment.

> Basic Information
>
> Sequence length constraints
>
> The 3' model expects an input sequence which is 300bp long (stop codon + 297). It will handle shorter sequences (although < 11 cannot be masked) and in theory can even predict up to 512 - but this is out-of-distribution and likely performs very poorly as the positional encodings are not adapted for this.
>
> The 5' model expects an input which is 1003bp long (1000 + start codon). Longer sequences will not work, shorter sequences must be padded (e.g. with a fixed sequence), otherwise the start codon gets the wrong positional encoding which confuses the model.

I am looking to use the model on some 5’ sequences that have < 1003 bp. When you say “"shorter sequences must be padded with (e.g. with a fixed sequence)” what do you mean exactly? Choose any of “A”, “T”, “C”, or “G”? Does this affect the model at all? How did you all train for sequences like this? I would like to follow whatever you did in your model development.

I’ve also considered padding with “N” which would give the token [UNK].

Thanks for your help.

**Comment by Karollus** [comment]
We only encountered this issue in one dataset, which was an MPRA where different promoters were tested in a fixed plasmid context. There we just used this fixed plasmid context to get the necessary padding up to a length of 1003, and it gave decent results (better than unpadded, at least). However, I have to say that we did not systematically evaluate this as it generally wasn't necessary for our use-cases.

So in case your sequences are also tested in some fixed context, you could use that context to pad. If not, then [PAD] tokens could work (I believe they will be masked by default in the attention, so they should not contribute to the embeddings), However, if your sequences vary heavily in length, this may nevertheless create problems. Unfortunately, I am not sure I have a good solution for that, maybe @johahi has an idea?

**Comment by Mjvolk3** [comment]
## S288c Genes Without 1003 bp Upstream

```python
not_1003_upstream = ['YAL068W-A', 'YAL069W', 'YBL113W-A', 'YCR108C', 'YDR545C-A', 'YEL077W-A', 'YER190C-B', 'YFL067W', 'YFL068W', 'YGR296C-B', 'YHL050W-A', 'YHR219C-A', 'YIL177W-A', 'YJL225W-A', 'YJR162C', 'YKL225W', 'YLL067W-A', 'YLR467C-A', 'YML133W-B', 'YMR326C', 'YNL339W-B', 'YNR077C', 'YOL166W-A', 'YOR396C-A']
```

Most of these genes are the edge of a chromosome.

![image](https://github.com/gagneurlab/SpeciesLM/assets/48566106/8ffbec48-ff59-4714-8f90-6c884ebc1465)

### SpeciesLM Five Prime S288c Genes Less Than 1003 bp

`dataset_five_prime` reads "SpeciesLM/data/Sequences/Annotation/Assembled/five_prime.parquet".

```python
>>> print({i:len(dataset_five_prime.loc[dataset_five_prime['gene_id'] == i, "five_prime_seq"].iloc[0]) for i in not_1003_upstream})
{'YAL068W-A': 540
'YAL069W': 337
'YBL113W-A': 648
'YCR108C': 435
'YDR545C-A': 594
'YEL077W-A': 632
'YER190C-B': 715
'YFL067W': 838
'YFL068W': 55
'YGR296C-B': 715
'YHL050W-A': 813
'YHR219C-A': 965
'YIL177W-A': 848
'YJL225W-A': 831
'YJR162C': 493
'YKL225W': 453
'YLL067W-A': 845
'YLR467C-A': 648
'YML133W-B': 829
'YMR326C': 632
'YNL339W-B': 736
'YNR077C': 795
'YOL166W-A': 587
'YOR396C-A': 797}
```

How did you deal with these genes?

**Comment by johahi** [comment]
Hi Michael, thanks for your interest! 
So for now we throw these out during training, as these are only very few sequences. 
Regarding padding shorter sequences during eval, I guess left-padding them using PAD tokens (to ensure anchoring with the start codon) and supplying the correct attention mask makes most sense, but we haven't tried this yet. Note that your embedding then needs to be pooled over the shorter seq as well...

**Comment by Mjvolk3** [comment]
Of course, it is very exciting work. I am a bit confused on the pooling part, and I'm a bit new working with language models so maybe you could help me clarify. When I inspect the hidden state size I get this

```python
outputs.hidden_states[-1].shape
torch.Size([1, 1001, 768])
```
Typically I'll take `(8,)` for the hidden states as shown in [ModelUsage.py](https://github.com/gagneurlab/SpeciesLM/blob/main/ModelUsage.ipynb). Then I will `mean` over the hidden states and over dim 1 (dim with 1001). Are you suggesting that I would need to slice this dimension according to sequence length then take the mean over this slice? If so, how do you slice the 1001 length since it is not 1003?

Just out of curiosity I looked at the downstream model and see this dimension is also `sequence_length - 2`.
```python
outputs.hidden_states[-1].shape
torch.Size([1, 298, 768])
```

Thanks any help.

**Comment by Mjvolk3** [comment]
When the sequences are `<1003 bp` the 1 dim shown before is `1001`, I believe this must be because the start codon is given a single token, so the length of tokens always ends up being `sequence_length - 2`. For now I just left pad with `0` until hitting length `1001` and keep track of the index which was left padded. Then when I am going to take the mean over this dimension that is now `1001` since padding, I slice it according to the length of pad. @johahi I think this was what you meant in your response.

Is it any concern that `'input_ids'`, `'token_type_ids'`, and `'attention_mask'` all get left padded by 0's? I assume this is good, so the attention mask is not on the padded tokens. Thanks again.

**Comment by johahi** [comment]
The reason for the model outputs being shorter than the input sequence is the overlapping 6-mer tokenization. It reduces a sequence of length n to n-5 tokens (e.g., ATGCCCC -> ATGCCC TGCCCC). Since the model has a start, end and a species token, you end up at n-5+3 = n-2 tokens.

Instead of left padding with 0, you could left pad only after the species token, so for instance:
[2,4424,0,0,0,64,...,3] where 2 is CLS, 4424 is a species token, 64 the beginning of your seq, 3 is EOS. For the embedding, you'd have to remove the padding again then.
The model should probably not attend the padded positions (attention_mask), token_type_ids is a left-over from the original BERT and not relevant for Species LM.
